src/cache.c

```diff
@@ -393,21 +474,63 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
     hlist_add_head(&new_entry->ht_list, &cache->map->ht_list_head[HASH(key)]);
 
     CACHE_ASSERT(cache);
+#else
+    assert(cache->list_size <= cache->capacity);
+    /* Before adding new element to cach, we should check the status
+     * of cache.
+     */
+    if (cache->list_size == cache->capacity) {
+        for (int i = 0; i < 1000; i++) {
+            if (!list_empty(cache->lists[i])) {
+                lfu_entry_t *delete_target =
+                    list_last_entry(cache->lists[i], lfu_entry_t, list);
+                list_del_init(&delete_target->list);
+                hlist_del_init(&delete_target->ht_list);
+                delete_value = delete_target->value;
+                cache->list_size--;
+                mpool_free(cache_mp, delete_target);
+                break;
+            }
+        }
+    }
+    lfu_entry_t *new_entry = mpool_alloc(cache_mp);
+    new_entry->key = key;
+    new_entry->value = value;
+    new_entry->frequency = 0;
+    list_add(&new_entry->list, cache->lists[new_entry->frequency++]);
+    cache->list_size++;
+    hlist_add_head(&new_entry->ht_list, &cache->map->ht_list_head[HASH(key)]);
+    assert(cache->list_size <= cache->capacity);
+#endif
     return delete_value;
 }
 
 void cache_free(cache_t *cache, void (*callback)(void *))
 {
+#if RV32_HAS(ARC)
     for (int i = 0; i < N_CACHE_LIST_TYPES; i++) {
         arc_entry_t *entry, *safe;
 #ifdef __HAVE_TYPEOF
         list_for_each_entry_safe (entry, safe, cache->lists[i], list)
 #else
         list_for_each_entry_safe (entry, safe, cache->lists[i], list,
                                   arc_entry_t)
+#endif
+#else
+    for (int i = 0; i < 1000; i++) {
```

Comment:
Why `1000`?

Reply:
This would modify with threshold we set, maybe I define a marco as threshold.